pkg/querier/querier.go

```diff
@@ -191,28 +202,86 @@ func (q querier) Select(sp *storage.SelectParams, matchers ...*labels.Matcher) (
 		}
 	}
 
-	return storage.NewMergeSeriesSet(result, nil), nil, nil
+	// we have all the sets from different sources (chunk from store, chunks from ingesters,
+	// time series from store and time series from ingesters).
+	return q.mergeSeriesSets(result), nil, nil
 }
 
 // LabelsValue implements storage.Querier.
 func (q querier) LabelValues(name string) ([]string, storage.Warnings, error) {
-	lv, err := q.distributor.LabelValuesForLabelName(q.ctx, model.LabelName(name))
-	return lv, nil, err
+	return q.primary.LabelValues(name)
 }
 
 func (q querier) LabelNames() ([]string, storage.Warnings, error) {
-	ln, err := q.distributor.LabelNames(q.ctx)
-	return ln, nil, err
+	return q.primary.LabelNames()
 }
 
-func (q querier) metadataQuery(matchers ...*labels.Matcher) (storage.SeriesSet, storage.Warnings, error) {
-	ms, err := q.distributor.MetricsForLabelMatchers(q.ctx, model.Time(q.mint), model.Time(q.maxt), matchers...)
-	if err != nil {
-		return nil, nil, err
+func (querier) Close() error {
+	return nil
+}
+
+func (q querier) mergeSeriesSets(sets []storage.SeriesSet) storage.SeriesSet {
+	// Here we deal with sets that are based on chunks and build single set from them.
+	// Remaining sets are merged with chunks-based one using storage.NewMergeSeriesSet
+
+	otherSets := []storage.SeriesSet(nil)
+	chunks := []chunk.Chunk(nil)
+
+	for _, set := range sets {
+		if !set.Next() {
+			// nothing in this set
+			continue
+		}
+
+		s := set.At()
+		if sc, ok := s.(SeriesWithChunks); ok {
+			chunks = append(chunks, sc.Chunks()...)
+
+			// iterate over remaining series in this set, and store chunks
+			// Here we assume that all remaining series in the set are also backed-up by chunks.
+			// If not, there will be panics.
+			for set.Next() {
```

Comment:
Could you confirm me this is the way it's working right now (before this PR), and the configured `iteratorFunc` doesn't affect this, please?

Reply:
`iteratorFunc` has somewhat unfortunate name, because what it is really doing is processing different chunks that belong to the same series (this is important) into single iterator. So it's only handling single series at the time.

This code on the other hand, is iterating over the series in the set, and simply collecting chunks. They are sorted into per-series chunks by using `partitionChunks` function.

Previously this step happened in `unifiedChunkQuerier.Select` method, which dealt with chunk-based stores (chunks from all stores were returned by its `Get` method). This new method deals with both chunks-based and non-chunks-based sets.